src/main/java/com/google/summit/ast/declaration/FieldDeclaration.kt

```diff
@@ -18,24 +18,64 @@ package com.google.summit.ast.declaration
 
 import com.google.summit.ast.Identifier
 import com.google.summit.ast.Node
+import com.google.summit.ast.NodeWithSourceLocation
 import com.google.summit.ast.SourceLocation
 import com.google.summit.ast.TypeRef
 import com.google.summit.ast.expression.Expression
+import com.google.summit.ast.modifier.Modifiable
+import com.google.summit.ast.modifier.Modifier
+
+/**
+ * Interface that is common between [FieldDeclarationGroup] and [FieldDeclaration]. Includes
```

Comment:
This doesn't pass my smell test for needing to be its own type: does it avoid non-trivial code duplication? is it an interface that's actually useful to manipulate its implementations abstractly?

With respect to the latter point, I struggle to imagine any analysis or computation based on (1) modifiers (2) type... nothing else.

Reply:
> is it an interface that's actually useful to manipulate its implementations abstractly?

Nope, and that wasn't what I intended it to be used for, which is why it is `private`.

The way I see it, there are now two ways to delegate properties in nodes:
1. Manual
```kotlin
class FieldDeclaration : Declaration() {
  var modifiers: List<Modifier>
    // ...
}
```
2. Implement & override interface (`Modifiable` or `Field`)
```kotlin
class FieldDeclaration : Declaration(), Modifiable {
  override var modifiers: List<Modifier>
    // ...
}
```

When I first wanted to delegate `type` also, I manually added a getter property, but I considered that it might be better for consistency if it was delegated using an interface too, so I made `Field` that contains all the properties that will be delegated from `FieldDeclaration` to `FieldDeclarationGroup`.

To be clear, I'm not sure about this interface either, but I hope you see the motivation behind it. Thoughts?

Reply:
You inspired me to go back and think about whether `Modifiable` should be factored out into a type. I think it clears the first bar: avoiding code duplication of a number of helper functions. That said, could it serve the same purpose as a member of (`FieldDeclarationGroup` and `ModifiableDeclaration`) rather than an implemented interface?

Back to `Field`...

I understand your argument to be consistency in delegating through an interface. TBH, the benefits seems less clear to me than factoring out `Modifiable`. That said, given that it is package-private, my concerns about minimizing inheritance and avoiding fragility are mostly moot. So no blocking objection.

Reply:
> You inspired me to go back and think about whether `Modifiable` should be factored out into a type. I think it clears the first bar: avoiding code duplication of a number of helper functions. That said, could it serve the same purpose as a member of (`FieldDeclarationGroup` and `ModifiableDeclaration`) rather than an implemented interface?

Could you elaborate on how the composition design would look like, and what its benefit is?

Reply:
Roughly:
```
class Modifiers {
  var all: List<Modifier>
  // ... various helper functions
}

class FieldDeclarationGroup {
  val modifiers: Modifiers
}

class ModifiableDeclaration {
  val modifiers: Modifiers
}
```
I will send you some links in chat

Reply:
I see how this could work from the client-side (especially `modifiers` being called `all` avoids a situation like `node.modifiers.modifiers`), but it's still unclear to me why this design would be beneficial over the current one, especially when it doesn't seem like the current design is a misuse of inheritance.

Additionally, the interface for clients would be less convenient due to an additional field access (e.g. `node.modifiers.hasKeyword` vs `node.hasKeyword`). There would also be no way to check whether a `Node` has-a `Modifiable` (correct me if I'm wrong) vs. is-a, which does occur at least once already.

P.S. Item 18 states,
> The problems discussed in this item do not apply to _interface inheritance_ (when a class implements an interface...

Reply:
> it doesn't seem like the current design is a misuse of inheritance

Certainly your code is functionally correct as-is. This is a pattern to ensure maintainability as it evolves over time.

Yup, those are disadvantages. The first one is minor. The second one bites a little more, and you'd have to handle the two cases.

Re: P.S. I believe this qualification needs to be be removed for Kotlin and/or interface methods having default implementations (which are basically abstract classes).

--- 

I think we've hit on the trade-offs, and I hope the arguments for composition at least make sense. Your code choice. I've already approved.